dask_match/core.py

```diff
@@ -247,13 +250,26 @@ def dtypes(self):
     def _meta(self):
         raise NotImplementedError()
 
+    def _layer(self):
+        if isinstance(self, Fusable):
+            return _subgraph_callable_layer(
+                self._name,
+                self._block_subgraph(),
+                self._subgraph_dependencies(),
+                self.npartitions,
+            )
+        else:
+            raise NotImplementedError
+
     def __dask_graph__(self):
         """Traverse expression tree, collect layers"""
+
         stack = [self]
         seen = set()
         layers = []
+        fusable = _fusable_ops(self)
         while stack:
-            expr = stack.pop()
+            expr = _fuse_expr_groups(stack.pop(), fusable)
```

Comment:
I don't think that this concept should be embedded directly into the Expr class.  This feels like a specific abstraction finding its way into what should be very general (the US Tax code is a good example of this).

Instead, I would expect us to have a pass that crawled over the expression tree, found groups that could be fused into a single Blockwise object, and then did that, returning a new expression tree with fewer nodes in it.

Reply:
>Instead, I would expect us to have a pass that crawled over the expression tree, found groups that could be fused into a single Blockwise object, and then did that, returning a new expression tree with fewer nodes in it.

Yes. My original plan was to actually fuse `Blockwise` expressions into new `Blockwise` expressions like we do in for `Blockwise` layers in dask/dask. However, I soon realized that I actually **dislike** that we need to do this in dask/dask.  The opaque composite layers we end up crating tend to loose most of their value/meaning after fusion (besides their ability to ensure that the appropriate "low-level" tasks get fused).

My mind is still very open on the subject (and is very likely to change), because I do appreciate the logical simplicity of treating fusion as a distinct `Expr` optimization pass. With that in mind, I currently like the idea of only applying fusion at graph-materialization time.